red4ext-sys/src/interop.rs

```diff
@@ -30,11 +30,34 @@ pub struct CName {
     hash: u64,
 }
 
+impl From<u64> for CName {
+    fn from(hash: u64) -> Self {
+        Self { hash }
+    }
+}
+
 impl CName {
     #[inline]
     pub const fn new(str: &str) -> Self {
         Self { hash: fnv1a64(str) }
     }
+
+    pub const fn to_u64(self) -> u64 {
+        self.hash
+    }
+
+    pub fn add(str: &str) -> Self {
+        if str.is_empty() {
+            return Self { hash: 0 };
+        }
+        let cname = Self::new(str);
+        if crate::ffi::exists_cname(&cname) {
+            return cname;
+        }
+        let created = crate::ffi::add_cname(str);
+        assert_eq!(created, cname);
+        created
+    }
```

Comment:
I'd probably just include this in `CName::new`, it's a change in behavior but I don't think there are any valid uses cases for `CName`s that aren't backed by anything

Reply:
might make sense to include `CName::new_unpooled` for tests and things that dont interact with the game

Reply:
This one is problematic since `CName::new` is `const`.

Inside function's block, string slice equality check could be provided by nightly (not great), by a 3rd-party crate like [eq_str](https://docs.rs/const_cmp/latest/const_cmp/fn.eq_str.html) (most likely be preferred) or hand-crafted.

But I don't think the `ffi` methods can or should (please correct me if I'm wrong).

Reply:
hmm you're right, and it's pretty useful it have a `const` function,
it might make sense to introduce something like this
```
    pub fn ensure_pooled(&self) {
        if str.is_valid() {
            return;
        }
        crate::ffi::add_cname(str);
    }
```

Reply:
instead of the `add` method

Reply:
hm actually, nevermind we don't have the `str` at this point

Reply:
I think this should just be defined like this then:
```
pub fn new_pooled(str: &str) -> Self
```

Reply:
Also current `CName::new` generates hash without checking provided `&str` is neither empty nor `None` (as [specified in RED4ext.SDK](https://github.com/WopsS/RED4ext.SDK/blob/d017563eb4cb5d875a49e47fe21615bfeee00e1d/include/RED4ext/CName.hpp#L84)), so depending if we want to be pedantic we could address it :

- `assert_eq!(CName::new("").hash, 0);`
- `assert_eq!(CName::new("None").hash, 0);`

Reply:
I can't see an empty string check in red4ext, it seems it'd generate a hash of `0xCBF29CE484222325` for an empty string
we can add a check for `None` though

Reply:
Ok I added the check for `None` directly in `CName::new` using `const_str`, and renamed `add` method as `new_pooled`.
The empty string will still generate a hash of `0xCBF29CE484222325`, as per RED4ext.SDK.